wallet: upgradewallet fixes, improvements, test coverage

jonatack opened the pull request:
This follows up on #18836 and #20282 to fix and improve the as-yet unreleased `upgradewallet` feature and also implement review follow-up in https://github.com/bitcoin/bitcoin/pull/18836#discussion_r519328607.

This PR fixes 4 upgradewallet issues:

- this bug: https://github.com/bitcoin/bitcoin/pull/20403#discussion_r526063920
- it returns nothing in the absence of an RPC error, which isn't reassuring for users
- it returns the same thing both in the case of a successful upgrade and when no upgrade took place
- the error message object is currently dead code

This PR fixes the above and provides:

...user feedback to not silently return without upgrading
```
{
  "wallet_name": "disable private keys",
  "previous_version": 169900,
  "current_version": 169900,
  "result": "Already at latest version. Wallet version unchanged."
}
```
...better feedback after successfully upgrading
```
{
  "wallet_name": "watch-only",
  "previous_version": 159900,
  "current_version": 169900,
  "result": "Wallet upgraded successfully from version 159900 to version 169900."
}
```
...helpful error responses
```
{
  "wallet_name": "blank",
  "previous_version": 169900,
  "current_version": 169900,
  "error": "Cannot downgrade wallet from version 169900 to version 159900. Wallet version unchanged."
}
{
  "wallet_name": "blank",
  "previous_version": 130000,
  "current_version": 130000,
  "error": "Cannot upgrade a non HD split wallet from version 130000 to version 169899 without upgrading to support pre-split keypool. Please use version 169900 or no version specified."
}
```
updated help:
```
upgradewallet ( version )

Upgrade the wallet. Upgrades to the latest version if no version number is specified.
New keys may be generated and a new wallet backup will need to be made.
Arguments:
1. version    (numeric, optional, default=169900) The version number to upgrade to. Default is the latest wallet version.

Result:
{                            (json object)
  "wallet_name" : "str",     (string) Name of wallet this operation was performed on
  "previous_version" : n,    (numeric) Version of wallet before this operation
  "current_version" : n,     (numeric) Version of wallet after this operation
  "result" : "str",          (string, optional) Description of result, if no error
  "error" : "str"            (string, optional) Error message (if there is one)
}
```

jonatack committed:
wallet: fix and improve upgradewallet result responses

jonatack committed:
wallet: refactor GetClosestWalletFeature()

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4126,7 +4126,9 @@ bool CWallet::UpgradeWallet(int version, bilingual_str& error, std::vector<bilin
-    SetMinVersion(GetClosestWalletFeature(version));
+    const WalletFeature new_version{GetClosestWalletFeature(version)};
+    SetMinVersion(new_version);
+    version = static_cast<int>(new_version); // return actual updated version
```
Is `static_cast` necessary here? `WalletFeature` should already be an int.

jonatack replied:
Oh you're right, looks like we can just do `version = new_version` here. Will test a push without the cast. Thanks!

achow101 commented:
Code review ACK b95925c751fb97a1a17bf5a540ec06f0634ea72e

jonatack commented:
Removed the cast in UpgradeWallet() per `git diff b95925c 377fa31` (thanks!)
```diff
@@ -4128,7 +4128,7 @@ bool CWallet::UpgradeWallet(int& version, bilingual_str& error, std::vector<bili
     // Permanently upgrade to the version
     const WalletFeature new_version{GetClosestWalletFeature(version)};
     SetMinVersion(new_version);
-    version = static_cast<int>(new_version); // return actual updated version
+    version = new_version; // return actual updated version
```

achow101 commented:
ACK 377fa31

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4475,12 +4479,26 @@ static RPCHelpMan upgradewallet()
+        obj.pushKV("result", result);
+    }
     if (!error.empty()) {
         obj.pushKV("error", error.original);
```
It looks like this is dead code, and always has been.

It would be easier to have one way to deal with errors. Either return them as `JSONRPCError` or as `{"error":...}`, but not both. Also, the error approach should be documented.

jonatack replied:
Yes, I plan to have a look at this later today. That's the second part of my review feedback from #20282.

In #20391 I proposed returning an error field rather than raising an RPC error for this type of errors but wasn't sure which is preferred.

MarcoFalke replied:
I think either is fine as long as it is documented and consistent. An RPCError makes sense when the return object would otherwise be empty and would have to omit a lot of optional fields, except for the error.

jonatack replied:
Proposing this:
```
$ ./src/bitcoin-cli -signet -rpcwallet="blank" upgradewallet 5
{
  "wallet_name": "blank",
  "previous_version": 169900,
  "current_version": 169900,
  "error": "Cannot downgrade wallet from version 169900 to version 5"
}
```

jonatack replied:
(currently that returns):
```
error code: -4
error message:
Cannot downgrade wallet
```

jonatack replied:
Done

jonatack replied:
Let me know if there is anything needed WRT documenting the error approach.

jonatack committed:
wallet: fix and improve upgradewallet error responses

jonatack commented:
Rebased after the merge of #20139 and addressed https://github.com/bitcoin/bitcoin/pull/20403#discussion_r524992713 with 877aba836ec48bfa881ec0dd90299015a3c27cce.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4474,11 +4478,25 @@ static RPCHelpMan upgradewallet()
+    if (!result.empty()) {
+        obj.pushKV("result", result);
+    }
     if (!error.empty()) {
```
```suggestion
    else { CHECK_NONFATAL(!error.empty());
```

jonatack replied:
Thanks--done.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -4113,20 +4113,25 @@ bool CWallet::UpgradeWallet(int version, bilingual_str& error)
     // Do not upgrade versions to any version between HD_SPLIT and FEATURE_PRE_SPLIT_KEYPOOL unless already supporting HD_SPLIT
     if (!CanSupportFeature(FEATURE_HD_SPLIT) && version >= FEATURE_HD_SPLIT && version < FEATURE_PRE_SPLIT_KEYPOOL) {
-        error = _("Cannot upgrade a non HD split wallet without upgrading to support pre split keypool. Please use version 169900 or no version specified.");
+        error = strprintf(_("Cannot upgrade a non HD split wallet from version %i to version %i without upgrading to support pre-split keypool. Please use version %i or no version specified."),
```
the translation shouldn't be changed here, if you want this to be backported

jonatack replied:
Oh right (thanks!) and same for the translation just above as well, I suppose
```diff
-        error = _("Cannot downgrade wallet");
+        error = strprintf(_("Cannot downgrade wallet from version %i to version %i. Wallet version unchanged."), prev_version, version);
```

MarcoFalke replied:
If you like the translation changes, you can put them into a separate commit and write "not for backport" in the commit body

jonatack replied:
Done, thanks

jonatack replied:
Thought: do translated strings need to be static? I should look at the implementation.

MarcoFalke replied:
They can only be literals, but may include format specifiers

jonatack committed:
wallet (not for backport): improve upgradewallet error messages

MarcoFalke commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -158,40 +170,29 @@ def copy_split_hd():
-        # upgraded wallet version should be greater than older one
-        assert_greater_than(new_version, old_version)
+        self.log.info("Test upgradewallet without version arguments")
+        self.test_upgradewallet(wallet, previous_version=159900, requested_version=169900)
```
This is not testing what the log claims to test

jonatack replied:
Good catch

jonatack replied:
Fixed

jonatack commented:
Found an edge case bug in the first commit, "wallet: enable CWallet::UpgradeWallet to return updated version", and the fix led to a simpler implementation.

Fixed two tests in the third commit, "wallet: fix and improve upgradewallet result responses"; thanks @MarcoFalke for the catching one of them. The other one led to seeing the edge case bug.

Re-verified build and wallet_upgradewallet test at each commit.

<details><summary>Diff: <code>git diff dfec0a8 25909e6</code></summary><p>

```diff
jon@purity:~/projects/bitcoin/bitcoin (upgradewallet-improvements)$ git diff dfec0a8 25909e6
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 17631c58e7..10a7a07ee7 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4127,15 +4127,15 @@ bool CWallet::UpgradeWallet(int& version, bilingual_str& error)
     }
 
     // Permanently upgrade to the version
-    const WalletFeature new_version{GetClosestWalletFeature(version)};
-    SetMinVersion(new_version);
-    version = new_version; // return actual updated version
+    SetMinVersion(GetClosestWalletFeature(version));
 
     for (auto spk_man : GetActiveScriptPubKeyMans()) {
         if (!spk_man->Upgrade(prev_version, version, error)) {
             return false;
         }
     }
+
+    version = GetVersion(); // return actual updated version
     return true;
 }
 
diff --git a/test/functional/wallet_upgradewallet.py b/test/functional/wallet_upgradewallet.py
index cec3dd2b6d..7cae913fa1 100755
--- a/test/functional/wallet_upgradewallet.py
+++ b/test/functional/wallet_upgradewallet.py
@@ -90,10 +90,10 @@ class UpgradeWalletTest(BitcoinTestFramework):
             v16_3_node.submitblock(b)
         assert_equal(v16_3_node.getblockcount(), to_height)
 
-    def test_upgradewallet(self, wallet, previous_version, requested_version, unchanged=False):
-        new_version = previous_version if unchanged else requested_version
+    def test_upgradewallet(self, wallet, previous_version, requested_version, expected_version=None, no_arg=False, unchanged=False):
+        new_version = previous_version if unchanged else expected_version if expected_version else requested_version
         assert_equal(wallet.getwalletinfo()["walletversion"], previous_version)
-        assert_equal(wallet.upgradewallet(requested_version),
+        assert_equal(wallet.upgradewallet() if no_arg else wallet.upgradewallet(requested_version),
             {
                 "wallet_name": "",
                 "previous_version": previous_version,
@@ -182,7 +182,7 @@ class UpgradeWalletTest(BitcoinTestFramework):
         copy_v16()
         wallet = node_master.get_wallet_rpc(self.default_wallet_name)
         self.log.info("Test upgradewallet without version arguments")
-        self.test_upgradewallet(wallet, previous_version=159900, requested_version=169900)
+        self.test_upgradewallet(wallet, previous_version=159900, requested_version=169900, no_arg=True)
         # wallet should still contain the same balance
         assert_equal(wallet.getbalance(), v16_3_balance)
 
@@ -342,7 +342,7 @@ class UpgradeWalletTest(BitcoinTestFramework):
         old_kvs = dump_bdb_kv(node_master_wallet)
         defaultkey = old_kvs[b'\x0adefaultkey']
         self.log.info("Upgrade the wallet. Should still have the same default key.")
-        self.test_upgradewallet(wallet, previous_version=139900, requested_version=169900)
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=169900)
         new_kvs = dump_bdb_kv(node_master_wallet)
         up_defaultkey = new_kvs[b'\x0adefaultkey']
         assert_equal(defaultkey, up_defaultkey)
```
</p></details>

jonatack commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -333,14 +341,15 @@ def copy_split_hd():
-        # Upgrade the wallet. Should still have the same default key
-        wallet.upgradewallet(159900)
+        self.log.info("Upgrade the wallet. Should still have the same default key.")
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=169900)
```
@achow101 if current version is 139900, upgradewallet 159900 is called, is the wallet actually upgrading to 169900 the expected behavior ?

achow101 replied:
Versions less than `FEATURE_HD_SPLIT` cannot upgrade to `FEATURE_HD_SPLIT` or `FEATURE_NO_DEFAULT_KEY`. They must upgrade to `FEATURE_PRE_SPLIT_KEYPOOL` at which time both the previous features will also be applied.

If a user specifies `FEATURE_NO_DEFAULT_KEY` on ` FEATURE_HD` wallet, we should probably give an error and tell the user they can't do that.

achow101 replied:
This test is incorrect. When `expected_version` is set, that is the version we are upgrading to. `requested_version` isn't used in that case.

achow101 replied:
Found it. Here's a fix.

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index d2e1be6402..7dbbf17302 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -453,7 +453,7 @@ bool LegacyScriptPubKeyMan::Upgrade(int prev_version, int new_version, bilingual
         hd_upgrade = true;
     }
     // Upgrade to HD chain split if necessary
-    if (IsFeatureSupported(new_version, FEATURE_HD_SPLIT)) {
+    if (!IsFeatureSupported(prev_version, FEATURE_HD_SPLIT) && IsFeatureSupported(new_version, FEATURE_HD_SPLIT)) {
         WalletLogPrintf("Upgrading wallet to use HD chain split\n");
         m_storage.SetMinVersion(FEATURE_PRE_SPLIT_KEYPOOL);
         split_upgrade = FEATURE_HD_SPLIT > prev_version;
diff --git a/test/functional/wallet_upgradewallet.py b/test/functional/wallet_upgradewallet.py
index 7cae913fa1..fdd52ba574 100755
--- a/test/functional/wallet_upgradewallet.py
+++ b/test/functional/wallet_upgradewallet.py
@@ -342,7 +342,7 @@ class UpgradeWalletTest(BitcoinTestFramework):
         old_kvs = dump_bdb_kv(node_master_wallet)
         defaultkey = old_kvs[b'\x0adefaultkey']
         self.log.info("Upgrade the wallet. Should still have the same default key.")
-        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=169900)
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=159900)
         new_kvs = dump_bdb_kv(node_master_wallet)
         up_defaultkey = new_kvs[b'\x0adefaultkey']
         assert_equal(defaultkey, up_defaultkey)
```

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -4133,6 +4133,8 @@ bool CWallet::UpgradeWallet(int version, bilingual_str& error)
         }
     }
+
+    version = GetVersion(); // return actual updated version
```
in commit 2310fe9685246b0d505495ee98d38c82a1730293: would be a smaller diff and less complexity if this was moved to the caller, no?

jonatack replied:
Yes, I tried removing the first commit and just calling GetVersion() in `RPCHelpMan upgradewallet()` but the first and last tests failed. Maybe a lock order/data race issue.

AssertionError: not({'wallet_name': '', 'previous_version': 159900, 'current_version': 169900, 'result': 'Wallet upgraded successfully from version 159900 to version 0.'} == {'wallet_name': '', 'previous_version': 159900, 'current_version': 169900, 'result': 'Wallet upgraded successfully from version 159900 to version 169900.'})

AssertionError: not({'wallet_name': '', 'previous_version': 139900, 'current_version': 169900, 'result': 'Wallet upgraded successfully from version 139900 to version 159900.'} == {'wallet_name': '', 'previous_version': 139900, 'current_version': 159900, 'result': 'Wallet upgraded successfully from version 139900 to version 159900.'})

jonatack replied:
Hm, will try again.

jonatack replied:
I didn't update one of the version variables the first time. Seems good.

jonatack commented:
Dropped the first commit to call `GetVersion()` in `RPCHelpMan upgradewallet()`.

achow101 committed:
Don't upgrade to HD split if it is already supported

It is unnecessary to upgrade to FEATURE_HD_SPLIT if this feature is
already supported by the wallet. Because upgrading to FEATURE_HD_SPLIT
actually requires upgrading to FEATURE_PRE_SPLIT_KEYPOOL, users would
accidentally be upgraded to FEATURE_PRE_SPLIT_KEYPOOL instead of nothing
being done.

Fixes the issue described at
https://github.com/bitcoin/bitcoin/pull/20403#discussion_r526063920

jonatack commented:
Pulled in #20420.

achow101 commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -158,40 +171,29 @@ def copy_split_hd():
-        wallet.upgradewallet(129999)
-        assert_equal(60000, wallet.getwalletinfo()['walletversion'])
+        # Test an "upgrade" from 60000 to 129999 has no effect, as next version is 130000
+        self.test_upgradewallet(wallet, previous_version=60000, requested_version=129900, expected_version=60000)
```
In commit 4dc34408bfe8fa1c229d99a3d0d03a42c83ced69 "wallet: fix and improve upgradewallet result responses"

`requested_version` should be 129999, not 129900

jonatack replied:
Good catch, thanks. Done.

achow101 commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -333,8 +332,8 @@ def copy_split_hd():
-        # Upgrade the wallet. Should still have the same default key
-        wallet.upgradewallet(159900)
+        self.log.info("Upgrade the wallet. Should still have the same default key.")
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900)
```
In commit 4dc34408bfe8fa1c229d99a3d0d03a42c83ced69 "wallet: fix and improve upgradewallet result responses"

The `assert_equal(wallet.getwalletinfo()["walletversion"], 159900)` below can be removed now.

jonatack replied:
Thanks for confirming that. Done.

achow101 commented:
ACK  3eb6f8b

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4214,7 +4214,7 @@ static RPCHelpMan sethdseed()
     // Do not do anything to non-HD wallets
     if (!pwallet->CanSupportFeature(FEATURE_HD)) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Cannot set a HD seed on a non-HD wallet. Use the upgradewallet RPC in order to upgrade a non-HD wallet to HD");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Cannot set an HD seed on a non-HD wallet. Use the upgradewallet RPC in order to upgrade a non-HD wallet to HD");
```
"a" is correct...

jonatack replied:
https://www.referencepointsoftware.com/a-vs-an-before-an-abbreviation/

MarcoFalke commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -92,6 +90,32 @@ def dumb_sync_blocks(self):
             v16_3_node.submitblock(b)
         assert_equal(v16_3_node.getblockcount(), to_height)
 
+    def test_upgradewallet(self, wallet, previous_version, requested_version=None, expected_version=None):
```
nit 99d56e357159c7154f69f28cb5587c5ca20d6594:

```suggestion
    def test_upgradewallet(self, wallet, *, previous_version, requested_version, expected_version=None):
```

Could force named args and pass requested_version in all cases for clarity. Only one call needs to be adjusted, I think.

jonatack replied:
> nit [99d56e3](https://github.com/bitcoin/bitcoin/commit/99d56e357159c7154f69f28cb5587c5ca20d6594):
> 
> Could force named args and pass requested_version in all cases for clarity. Only one call needs to be adjusted, I think.

Good idea.

jonatack replied:
> nit [99d56e3](https://github.com/bitcoin/bitcoin/commit/99d56e357159c7154f69f28cb5587c5ca20d6594):
> 
> Could force named args and pass requested_version in all cases for clarity. Only one call needs to be adjusted, I think.

Thanks! Good idea.

jonatack replied:
Done in e915a2a2c87dad5beb68a175b21a07136bcc21ca

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4475,11 +4479,27 @@ static RPCHelpMan upgradewallet()
+    if (!result.empty()) {
+        obj.pushKV("result", result);
+    } else {
+        CHECK_NONFATAL(!error.empty());
```
style nit ca8cd893bb56bf5d455154b0498b1f58f77d20ed:

This should check (before the if) that there is either an error or a result, but not both.

```cpp

CHECK_NONFATAL(error.empty != result.empty());

jonatack replied:
> style nit [ca8cd89](https://github.com/bitcoin/bitcoin/commit/ca8cd893bb56bf5d455154b0498b1f58f77d20ed):
> 
> This should check (before the if) that there is either an error or a result, but not both.

Good idea, that would be better.

jonatack replied:
Done in e915a2a2c87dad5beb68a175b21a07136bcc21ca

maflcko reviewed (approved):
review ACK 3eb6f8b2e61c24a22ea9396d86672307845f35eb 🛡

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 3eb6f8b2e61c24a22ea9396d86672307845f35eb 🛡
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUin9Av/YujicGdR4qxlxmUVaSijOicvW7fTqB6MnQH6+MJnzn5mApTxTVF4ZLZj
nrHGzoEVTcnok6tsDS62eGJvmuP/FxXDZExtb/DrjMdcXqjdP4kpH67tMGAnyrnw
V/M5psoZxs/u0qm4zFa08ni6OBaeSA3U8bZ5Qj6Uk1P9of3VblGEjlw8wa3P/JU0
71s8J3wcyfXY+GS5iQ1DIbz4SgP7Ch7wLPQtgvetEPRblTvhoAgueZeURmgwHai9
ZIf1T9DpMkkBuJiZM+BhygbYF5J3r7vD0hcT/rrp3XNHpLuuMK0/ijaoJk2VXC44
/pnjQNbNXW01NH2PiamTY84rdXJW6qt7NJQtuIeOm2jivvo+Av+DOboOR5KRURUW
m4Ux1i6VaJ1Hov34Sqcnm4zy+WEq2/Lkpl5viwSOqctm7Yak/R38X5wYAx8PmsLP
S8/HSAS4aZob64D/Kkakod5VG1xl9fFw7aUQVioBGQ17hQ3PBYyXnBaIfaOiMXyR
RaovSwTo
=l1/v
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1140e347b1496cc50a43fa6fca8080f5ac91dbec5a7d8191668b108f5c1bf70d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081140e347b1496cc50a43fa6fca8080f5ac91dbec5a7d8191668b108f5c1bf70df0103871354e509d8422a4de110ca94571d408fff010706983b688c47d1a40b98bd7391c43ee08f120bded4670fc478c105b679805a44642c2f3b86d34fa83cd8db3185781e40812cf08f1045fbe43ecf00820e0993aa03700540083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010840d27bbc3c6182e87c9c2882fe6473a08f12087b43e22ab3865430032e43750d4f3104fa6257688ff52ef7afa1693e885a89b08f1045fbe43ecf00824383ee24466dab90083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0109e4c19da3cee0115338e2fd21f0a428d08f1045fbe43f0f008072b913f048159f90083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f11b87d7b3eccedeae96a3c20440208c08f1207971a731feb40969f477971727bbab526e7bb95caa83d8997abf2ec9abc122dd08f1045fbe43ecf008c390c0e04140145b0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
Backported in #20490
